File: java/src/main/java/org/example/Data/MultipartUploadData.java
/*
* Copyright (c) 2021 PSPACE, inc. KSAN Development Team ksan@pspace.co.kr
* KSAN is a suite of free software: you can redistribute it and/or modify it under the terms of
* the GNU General Public License as published by the Free Software Foundation, either version
* 3 of the License. See LICENSE for details
*
* 본 프로그램 및 관련 소스코드, 문서 등 모든 자료는 있는 그대로 제공이 됩니다.
* KSAN 프로젝트의 개발자 및 개발사는 이 프로그램을 사용한 결과에 따른 어떠한 책임도 지지 않습니다.
* KSAN 개발팀은 사전 공지, 허락, 동의 없이 KSAN 개발에 관련된 모든 결과물에 대한 LICENSE 방식을 변경 할 권리가 있습니다.
*/
package org.example.Data;

import java.util.ArrayList;

import com.amazonaws.services.s3.model.PartETag;

public class MultipartUploadData {
	public String uploadId;
	public ArrayList<PartETag> parts;
	public StringBuilder body;
	public int partSize;

	public MultipartUploadData()
	{
		Init();
	}

	public void Init()
	{
		uploadId = "";
		body = new StringBuilder();
		parts = new ArrayList<PartETag>();
		partSize = 5 * MainData.MB;
	}

	public int nextPartNumber() { return parts.size() + 1; }
	public String getBody() { return body.toString(); }

	public void addPart(int partNumber, String eTag) { parts.add(new PartETag(partNumber, eTag)); }
	public void addPart(PartETag part) { parts.add(part); }
	public void appendBody(String data) { body.append(data); }
}

File: python/s3tests/data/multipart_upload_data.py
"""Multipart upload state ported from Java MultipartUploadV2Data."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from s3tests.data import main_data as md
from s3tests.utils import checksum


class MultipartUploadV2Data:
    def __init__(self, part_size: int = 5 * md.MB) -> None:
        self.upload_id = ""
        self.parts: List[Dict[str, Any]] = []
        self.body = ""
        self.part_size = part_size

    def next_part_number(self) -> int:
        return len(self.parts) + 1

    def get_body(self) -> str:
        return self.body

    def add_part(self, etag: str, part_number: Optional[int] = None) -> None:
        if part_number is None:
            part_number = self.next_part_number()
        self.parts.append({"ETag": etag, "PartNumber": part_number})

    def add_part_with_checksum(self, algorithm: str, part_response: Dict[str, Any]) -> None:
        part_number = self.next_part_number()
        part: Dict[str, Any] = {"PartNumber": part_number, "ETag": part_response["ETag"]}
        value = checksum.get_checksum(part_response, algorithm)
        if value is not None:
            checksum.set_checksum(part, algorithm, value)
        self.parts.append(part)

    def add_part_from_copy(self, algorithm: str, part_response: Dict[str, Any]) -> None:
        copy_result = part_response["CopyPartResult"]
        part_number = self.next_part_number()
        part: Dict[str, Any] = {"PartNumber": part_number, "ETag": copy_result["ETag"]}
        value = checksum.get_checksum(copy_result, algorithm)
        if value is not None:
            checksum.set_checksum(part, algorithm, value)
        self.parts.append(part)

    def append_body(self, data: str) -> None:
        self.body += data

    def completed_multipart_upload(self) -> Dict[str, Any]:
        return {"Parts": self.parts}


MultipartUploadData = MultipartUploadV2Data
